### PR TITLE
Specify APCu autoloader prefix

### LINKED ISSUE #7049: Reproducable build: deterministic APCU prefix
Each time when `composer install --no-dev --optimize-autoloader --apcu-autoloader` is run, a new random prefix is generated 

https://github.com/composer/composer/blob/be040f8e3111b38c1e49ae5cd6b929dbf5c39045/src/Composer/Autoload/AutoloadGenerator.php#L657

This is problematic if one tries to achieve reproducible builds where the same input leads to the same output. It would be cool if composer would offer an option to override the prefix or make it deterministic based on the `composer.lock` hash. Any other idea is welcome as well.

### PR DESCRIPTION
**Pull request by lstrojny**
Closes #7049 

Introduces new option `--apcu-autoloader-prefix` to `install`, `require`, `update`, `remove` and `--apcu-prefix` in `dump-autoload`.

### CODE REVIEW

File: src/Composer/Autoload/AutoloadGenerator.php [hunk 1]
```diff
@@ -858,7 +858,7 @@ public static function getLoader()
         }
 
         if ($this->apcu) {
-            $apcuPrefix = substr(base64_encode(md5(uniqid('', true), true)), 0, -3);
+            $apcuPrefix = is_string($this->apcu) ? $this->apcu : substr(base64_encode(md5(uniqid('', true), true)), 0, -3);
```
Comment (stof) [inline src/Composer/Autoload/AutoloadGenerator.php:861]: if we allow user input, we cannot assume that the content is fine to output in a string without any escaping. So we must use `var_export($value, true)` to generate the string representation instead (removing the quotes around the variables when using it in the code just below). Otherwise, we might generate invalid code.

File: src/Composer/Command/RemoveCommand.php [hunk 2]
```diff
@@ -54,7 +54,7 @@ protected function configure()
                 new InputOption('ignore-platform-reqs', null, InputOption::VALUE_NONE, 'Ignore all platform requirements (php & ext- packages).'),
                 new InputOption('optimize-autoloader', 'o', InputOption::VALUE_NONE, 'Optimize autoloader during autoloader dump'),
                 new InputOption('classmap-authoritative', 'a', InputOption::VALUE_NONE, 'Autoload classes from the classmap only. Implicitly enables `--optimize-autoloader`.'),
-                new InputOption('apcu-autoloader', null, InputOption::VALUE_NONE, 'Use APCu to cache found/not-found classes.'),
+                new InputOption('apcu-autoloader', null, InputOption::VALUE_OPTIONAL, 'Use APCu to cache found/not-found classes (passing a string sets the APCu prefix)'),
```
Comment (stof) [inline src/Composer/Command/RemoveCommand.php:57]: what happens when running `composer remove --apcu-autoloader some/package` ? Is `some/package` still treated as the argument or does it become the value of the option ?

Comment (lstrojny) [reply src/Composer/Command/RemoveCommand.php:57]: Uh, yeah, that’s indeed a problem and one that can probably only be solved by upgrading. What would work is a separate parameter with a required value, so `--apc-autoloader-prefix`. @Seldaek what do you think?

Comment (Seldaek) [reply src/Composer/Command/RemoveCommand.php:57]: I think this is not that important and can maybe wait until 2.1, because 2.0 should really be stabilized and shipped.

Comment (stof) [reply src/Composer/Command/RemoveCommand.php:57]: @lstrojny what is the min Symfony version in which this works fine ?

File: src/Composer/Command/DumpAutoloadCommand.php [hunk 3]
```diff
@@ -33,7 +33,7 @@ protected function configure()
                 new InputOption('no-scripts', null, InputOption::VALUE_NONE, 'Skips the execution of all scripts defined in composer.json file.'),
                 new InputOption('optimize', 'o', InputOption::VALUE_NONE, 'Optimizes PSR0 and PSR4 packages to be loaded with classmaps too, good for production.'),
                 new InputOption('classmap-authoritative', 'a', InputOption::VALUE_NONE, 'Autoload classes from the classmap only. Implicitly enables `--optimize`.'),
-                new InputOption('apcu', null, InputOption::VALUE_NONE, 'Use APCu to cache found/not-found classes.'),
+                new InputOption('apcu', null, InputOption::VALUE_OPTIONAL, 'Use APCu to cache found/not-found classes.'),
```
Comment (staabm) [inline src/Composer/Command/DumpAutoloadCommand.php:36]: Description should mention how to use with a string  argument

File: src/Composer/Command/RemoveCommand.php (hunk 2 above)
Comment (lstrojny) [reply src/Composer/Command/RemoveCommand.php:57]: @stof we would need at least this change https://github.com/symfony/symfony/commit/80867422acd83d2602187155e7eb978d6c3d0d44, which is earliest available in 3.3.0

But: in `RequireCommand` and `RemoveCommand` we would have both an option with an optional value (`apcu-autoloader`) as well as an argument with an optional value (`packages`). Assume `composer require --apcu-autoloader package/name` and the `apcu-autoloader` this will lead to `package/name` being treated as the value of the `apcu-autoloader` option. `composer require --apcu-autoloader -- package/name` would work as it clarifies what is meant. Doing so would be a BC break and unfortunately the worst one: a command that was previously working might silently change behavior.

In conclusion: we need to introduce a separate option because a newer Symfony version will only fix half of the problem to preserve backwards compatibility.

(My current stance would be that Symfony should probably forbid having options and arguments with optional values in the same command as it will automatically lead to a quite counter-intuitive user experience)

**Comment by lstrojny** [comment]
The current version of this PR introduces a separate option which is the only sensible path forward I see. Question: should passing an APCu autoloader prefix automatically turn on APCu autoloading?

**Comment by Seldaek** [comment]
Ok yeah probably makes sense the way it is. It's not like this flag will be needed by a ton of people anyway.. I'd say passing the prefix should turn on the apcu option too to avoid having to type this twice, because really setting the prefix if you are not using apcu makes no sense.

**Comment by lstrojny** [comment]
@Seldaek done and done

**Comment by lstrojny** [comment]
@Seldaek anything else needed?

**Comment by Seldaek** [comment]
Just free time to review one last time.

**Comment by Seldaek** [comment]
But it's in the right milestone, it won't get forgotten

**Comment by lstrojny** [comment]
Great

**Comment by Seldaek** [comment]
Thanks @lstrojny
